pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,514 @@
+const maxPods = 1024
+
+// Create a new map when we removed enough pods to avoid memory leaks
+// since Go maps never free memory.
+const mapReInitializeThreshold = 1000
+
+type userNsPodsManager interface {
+	getPodDir(podUID types.UID) string
+	listPodsFromDisk() ([]types.UID, error)
+}
+
+type usernsManager struct {
+	used     *bitArray
+	usedBy   map[string]uint32
+	removed  int
+	currPods int
+	kl       userNsPodsManager
+	sync.Mutex
+}
+
+// UserNamespace holds the configuration for the user namespace.
+type userNamespace struct {
+	// UIDs mappings for the user namespace.
+	UIDMappings []idMapping `json:"uidMappings"`
+	// GIDs mappings for the user namespace.
+	GIDMappings []idMapping `json:"gidMappings"`
+}
+
+// Pod user namespace mapping
+type idMapping struct {
+	// Required.
+	HostId uint32 `json:"hostId"`
+	// Required.
+	ContainerId uint32 `json:"containerId"`
+	// Required.
+	Length uint32 `json:"length"`
+}
+
+// mappingsFile is the file where the user namespace mappings are persisted.
+const mappingsFile = "userns"
+
+// writeMappingsToFile writes the specified user namespace configuration to the pod
+// directory.
+func (m *usernsManager) writeMappingsToFile(pod types.UID, userNs userNamespace) error {
+	dir := m.kl.getPodDir(pod)
+
+	data, err := json.Marshal(userNs)
+	if err != nil {
+		return err
+	}
+
+	fstore, err := utilstore.NewFileStore(dir, &utilfs.DefaultFs{})
+	if err != nil {
+		return err
+	}
+	if err := fstore.Write(mappingsFile, data); err != nil {
+		return err
+	}
+
+	// We need to fsync the parent dir so the file is guaranteed to be there.
+	// fstore guarantees an atomic write, we need durability too.
+	parentDir, err := os.Open(dir)
+	if err != nil {
+		return err
+	}
+
+	if err = parentDir.Sync(); err != nil {
+		// Ignore return here, there is already an error reported.
+		parentDir.Close()
+		return err
+	}
+
+	return parentDir.Close()
+}
+
+// readMappingsFromFile reads the user namespace configuration from the pod directory.
+func (m *usernsManager) readMappingsFromFile(pod types.UID) ([]byte, error) {
+	dir := m.kl.getPodDir(pod)
+	fstore, err := utilstore.NewFileStore(dir, &utilfs.DefaultFs{})
+	if err != nil {
+		return nil, err
+	}
+	return fstore.Read(mappingsFile)
+}
+
+func MakeUserNsManager(kl userNsPodsManager) (*usernsManager, error) {
+	m := usernsManager{
+		// Create a bitArray for all the UID space (2^32).
+		// As a by product of that, no index param to bitArray can be out of bounds (index is uint32).
+		used:   makeBitArray((math.MaxUint32 + 1) / userNsLength),
+		usedBy: make(map[string]uint32),
+		kl:     kl,
+	}
+	// First block is reserved for the host.
+	m.used.set(0)
+
+	// Second block will be used for phase II. Don't assign that range for now.
+	m.used.set(1)
+
+	// do not bother reading the list of pods if user namespaces are not enabled.
+	if !utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport) {
```

Comment:
Can this be moved further up?

Reply:
I think it is still better if the returned object marks the first and second blocks as set.

Reply:
It is simpler code for the tests too. We are doing an allocation and a set (that has no return value), so it seems reasonably safe.

What do you think @mrunalp ?

Reply:
Sounds good!